diff --git a/paper-server/src/main/java/com/destroystokyo/paper/Metrics.java b/paper-server/src/main/java/com/destroystokyo/paper/Metrics.java
index 9f4d73fd3..6993af920 100644
--- a/paper-server/src/main/java/com/destroystokyo/paper/Metrics.java
+++ b/paper-server/src/main/java/com/destroystokyo/paper/Metrics.java
@@ -676,6 +676,40 @@ public class Metrics {
 
                     return map;
                 }));
+
+                // Paper start - SIMD support - include in metrics
+                Map<String, Map<String, Integer>> simdSupportMap = new HashMap<>(2); // Empty until initialized
+                metrics.addCustomChart(new Metrics.DrilldownPie("simd_support", () -> {
+                    if (!gg.pufferfish.pufferfish.simd.SIMDDetection.isInitialized()) {
+                        return null;
+                    }
+                    if (simdSupportMap.isEmpty()) {
+                        // Initialize
+                        boolean isEnabled = gg.pufferfish.pufferfish.simd.SIMDDetection.isEnabled();
+
+                        // use details as lower dimension
+                        Map<String, Integer> entry = new HashMap<>(2);
+                        String details;
+                        if (isEnabled) {
+                            details = "int " + gg.pufferfish.pufferfish.simd.SIMDDetection.intVectorBitSize() + "*" + gg.pufferfish.pufferfish.simd.SIMDDetection.intElementSize() + ", float " + gg.pufferfish.pufferfish.simd.SIMDDetection.floatVectorBitSize() + "*" + gg.pufferfish.pufferfish.simd.SIMDDetection.floatElementSize();
+                        } else {
+                            if (!gg.pufferfish.pufferfish.simd.SIMDDetection.testRunCompleted()) {
+                                details = "test failed";
+                            } else if (gg.pufferfish.pufferfish.simd.SIMDDetection.unsupportingLaneSize()) {
+                                details = "no supporting lane size";
+                            } else {
+                                details = "other reason";
+                            }
+                        }
+                        entry.put(details, 1);
+
+                        // use enabled or not as higher dimension
+                        simdSupportMap.put(String.valueOf(isEnabled), entry);
+
+                    }
+                    return simdSupportMap;
+                }));
+                // Paper end - SIMD support - include in metrics
             }
 
         }
diff --git a/paper-server/src/main/java/org/spigotmc/AsyncCatcher.java b/paper-server/src/main/java/org/spigotmc/AsyncCatcher.java
index 7e7a0ed94..a8c759be4 100644
--- a/paper-server/src/main/java/org/spigotmc/AsyncCatcher.java
+++ b/paper-server/src/main/java/org/spigotmc/AsyncCatcher.java
@@ -7,7 +7,8 @@ public class AsyncCatcher {
     public static void catchOp(String reason) {
         if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThread()) { // Paper - chunk system
             MinecraftServer.LOGGER.error("Thread {} failed main thread check: {}", Thread.currentThread().getName(), reason, new Throwable()); // Paper
-            throw new IllegalStateException("Asynchronous " + reason + "!");
+            // Custom modification: Enhanced error message
+            throw new IllegalStateException("Asynchronous " + reason + "! This operation must be performed on the main thread.");
         }
     }
 }
